File: backend/internal/pkg/mailer/mailer.go
package mailer

import (
	"errors"
	"fmt"
	"log"
)

var ErrNotConfigured = errors.New("mailer not configured")

type Mailer interface {
	Send(to, subject, bodyText string) error
}

type ConsoleMailer struct{}

func (m *ConsoleMailer) Send(to, subject, bodyText string) error {
	log.Printf("[MAIL][console] to=%s subject=%s\n%s", to, subject, bodyText)
	return nil
}

type SMTPConfig struct {
	Host               string
	Port               int
	User               string
	Pass               string
	From               string
	StartTLS           bool
	InsecureSkipVerify bool
}

func NewFromSMTPConfig(cfg SMTPConfig) (Mailer, error) {
	if cfg.Host == "" || cfg.Port == 0 || cfg.From == "" {
		return nil, ErrNotConfigured
	}
	return NewSMTPMailer(cfg)
}

func FormatLoginOTPBody(code string, ttlMinutes int) string {
	return fmt.Sprintf(
		"Your CircuitWorks login code: %s\n\nThis code expires in %d minutes.\nIf you didn't request this, ignore this email.",
		code, ttlMinutes,
	)
}

File: back/utils/mailer.py
import smtplib
import logging
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from app.core.config import settings


async def send_otp_email(to_email: str, code: str, ttl_minutes: int):
    """Send OTP code via email"""
    if not settings.SMTP_HOST:
        # Development mode - log to console
        logging.info(f"[MAIL] To: {to_email}, Code: {code}, TTL: {ttl_minutes}")
        return
    
    subject = "Your CircuitWorks login code"
    body = f"""
    Your CircuitWorks login code: {code}
    
    This code expires in {ttl_minutes} minutes.
    If you didn't request this, ignore this email.
    """
    
    msg = MIMEMultipart()
    msg["From"] = settings.SMTP_FROM
    msg["To"] = to_email
    msg["Subject"] = subject
    msg.attach(MIMEText(body, "plain"))
    
    try:
        with smtplib.SMTP(settings.SMTP_HOST, settings.SMTP_PORT) as server:
            if settings.SMTP_STARTTLS:
                server.starttls()
            if settings.SMTP_USER and settings.SMTP_PASS:
                server.login(settings.SMTP_USER, settings.SMTP_PASS)
            server.send_message(msg)
    except Exception as e:
        logging.error(f"Failed to send email: {e}")
        raise